File: force-field/dart/vector.dart
import 'dart:math';

class Vector {
  double x;
  double y;
  
  Vector(this.x, this.y) {
  }
  
  Vector operator -(Vector v) {
    return new Vector(x - v.x, y - v.y);
  }
  
  Vector operator +(Vector v) {
    return new Vector(x + v.x, y + v.y);
  }
  
  Vector operator *(double d) {
    return new Vector(x * d, y * d);
  }
  
  void addInPlace(Vector v) {
    x += v.x;
    y += v.y;
  }
  
  double dot(Vector v) {
    return x * v.x + y * v.y;
  }
  
  Vector normalize() {
    double magnitude = sqrt(dot(this));
    if (magnitude == 0) return new Vector(0, 0);
    return this * (1.0 / magnitude);
  }
  
  String toString() {
    return "(${x}, ${y})";
  }
  
}

File: force-field/php/Vector.php
<?php

class Vector {
	public $x;
	public $y;

	public function __construct($x, $y)
	{
		$this->x = $x;
		$this->y = $y;
	}

	public function subtract($v)
	{
		return new static($this->x - $v->x, $this->y - $v->y);
	}

	public function add($v)
	{
		return new static($this->x + $v->x, $this->y + $v->y);
	}

	public function scale($d)
	{
		return new Vector($this->x * $d, $this->y * $d);
	}

	public function addInPlace($v)
	{
		$this->x += $v->x;
		$this->y += $v->y;
	}

	public function dot($v)
	{
		return ($this->x * $v->x) + ($this->y * $v->y);
	}

	public function normalize()
	{
		$magnitude = sqrt($this->dot($this));
		if ($magnitude == 0) return new static(0, 0);
		return $this->scale(1.0 / $magnitude);
	}

	public function __toString()
	{
		return sprintf('(%f, %f)', $this->x, $this->y);
	}
}
